Source/WebCore/platform/spiel/PlatformSpeechSynthesizerSpiel.cpp

```diff
@@ -0,0 +1,294 @@
+        if (isPaused)
+            self->m_platformSynthesizer.client().didPauseSpeaking(*self->m_utterance);
+        else
+            self->m_platformSynthesizer.client().didResumeSpeaking(*self->m_utterance);
+    }), this);
+
+    g_signal_connect_swapped(m_speaker.get(), "notify::voices", G_CALLBACK(+[](SpielSpeechWrapper* self, SpielSpeaker*) {
+        self->m_platformSynthesizer.client().voicesDidChange();
+    }), this);
+
+    m_speakerCreatedCallback();
+}
+
+SpielSpeechWrapper::~SpielSpeechWrapper()
+{
+    g_signal_handlers_disconnect_by_data(m_speaker.get(), this);
+}
+
+String SpielSpeechWrapper::generateVoiceURI(const GRefPtr<SpielVoice>& voice, const String& language)
+{
+    auto provider = spiel_voice_get_provider(voice.get());
+    return makeString(URI_PREFIX, span(spiel_provider_get_well_known_name(provider)), '#', span(spiel_voice_get_identifier(voice.get())), '#', language);
+}
+
+Vector<RefPtr<PlatformSpeechSynthesisVoice>> SpielSpeechWrapper::initializeVoiceList()
+{
+    Vector<RefPtr<PlatformSpeechSynthesisVoice>> platformVoices;
+    auto voices = spiel_speaker_get_voices(m_speaker.get());
+    unsigned position = 0;
+    m_voices.clear();
+    while (auto item = g_list_model_get_item(voices, position)) {
+        auto voice = SPIEL_VOICE(item);
+        auto name = makeString(span(spiel_voice_get_name(voice)));
+        const char* const* languages = spiel_voice_get_languages(voice);
+        for (unsigned i = 0; i < G_N_ELEMENTS(languages); i++) {
+            auto language = makeString(span(languages[i]));
+            bool isDefault = !i;
+            auto uri = generateVoiceURI(voice, language);
+            platformVoices.append(PlatformSpeechSynthesisVoice::create(uri, name, language, false, isDefault));
+            m_voices.add(uri, GRefPtr(voice));
+        }
+        position++;
+    }
+    return platformVoices;
+}
+
+void SpielSpeechWrapper::pause()
+{
+    if (!m_utterance)
+        return;
+
+    spiel_speaker_pause(m_speaker.get());
+}
+
+void SpielSpeechWrapper::resume()
+{
+    if (!m_utterance)
+        return;
+
+    spiel_speaker_resume(m_speaker.get());
+}
+
+void SpielSpeechWrapper::speakUtterance(RefPtr<PlatformSpeechSynthesisUtterance>&& utterance)
+{
+    ASSERT(!m_utterance);
+    ASSERT(utterance);
+    if (!utterance)
+        return;
+
+    if (!utterance->voice()) {
+        m_platformSynthesizer.client().didFinishSpeaking(*utterance);
+        return;
+    }
+
+    const auto& uri = utterance->voice()->voiceURI();
+    if (!m_voices.contains(uri)) {
+        WTFLogAlways("Unknown voice URI: %s", uri.utf8().data());
+        return;
+    }
+
+    // TODO: Detect whether the utterance text is XML and enable SSML if that is the case.
+    auto voice = m_voices.get(uri);
+    auto text = utterance->text().utf8();
+    auto spielUtterance = adoptGRef(spiel_utterance_new(text.data()));
+    auto language = utterance->lang().utf8();
```

Comment:
I would avoid `text` and `language` variables. Keep them if you feel like it but at least keep them as a String, the byproduct of calling `utf8()` gives me the shivers.